tests/Replication2/Helper/WaitForQueue.h

```diff
@@ -0,0 +1,101 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2021-2021 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Lars Maier
+////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+#include <map>
+#include <optional>
+#include <mutex>
+#include <variant>
+
+#include "Futures/Future.h"
+
+namespace arangodb::replication2::test {
+
+template<typename IndexType, typename ResultType>
+struct WaitForQueue {
+  WaitForQueue() = default;
+
+  static_assert(std::is_copy_constructible_v<ResultType>);
+
+  auto waitFor(IndexType index) -> futures::Future<ResultType> {
+    std::unique_lock guard(mutex);
+    if (index <= resolvedIndex) {
+      TRI_ASSERT(lastResult.has_value());
+      return futures::Future<ResultType>{*lastResult};
+    }
+
+    return queue.emplace(index, futures::Promise<ResultType>{})
+        ->second.getFuture();
+  }
+
+  void resolve(IndexType upTo, ResultType withValue) {
+    resolve(upTo, futures::Try<ResultType>{withValue});
+  }
+
+  void resolve(IndexType upTo, futures::Try<ResultType> withTry) {
+    auto resolveSet = QueueType{};
+    {
+      std::unique_lock guard(mutex);
+      for (auto it = queue.begin(); it != queue.end();) {
+        if (it->first > upTo) {
+          break;
+        }
+        auto node = queue.extract(it++);
+        resolveSet.insert(std::move(node));
+      }
+
+      resolvedIndex = upTo;
+      lastResult = withTry;
+    }
+
+    for (auto& [index, promise] : resolveSet) {
+      promise.setTry(std::move(withTry));
+    }
+  }
+
+  void resolveAll(ResultType withValue) {
+    resolveAll(futures::Try<ResultType>(withValue));
+  }
+  void resolveAll(futures::Try<ResultType> withTry) {
+    auto resolveSet = QueueType{};
+    {
+      std::unique_lock guard(mutex);
+      std::swap(resolveSet, queue);
+    }
```

Comment:
Should `resolvedIndex` and `lastResult` be updated?

Reply:
to which value should we update them?

Reply:
Similar to `resolve` above. `lastResult = withTry`, while `resolvedIndex` becomes the last index from `resolveSet`. This is more of a question than a suggestion, I'm not sure if this would do any good to the code.